crates/risc0/src/host/boundless.rs

```diff
@@ -0,0 +1,710 @@
+
+        // TODO: Can be done better?
+        let s3_url = read_env("BOUNDLESS_S3_URL")?;
+
+        let s3_use_presigned = read_env("BOUNDLESS_S3_NO_PRESIGNED")
+            .map(|_| false)
+            .unwrap_or(true);
+
+        // Upload the program(elf) to the boundless storage provider
+        let mut image_url = self.client.upload_program(&elf).await?;
+        tracing::info!("Image URL: {}", image_url);
+
+        // If we are not using presigned:
+        if !s3_use_presigned {
+            let s3_path = image_url.as_str().to_string().replace("s3://", "");
+            image_url = Url::parse(&format!("{s3_url}{s3_path}"))?;
+            tracing::info!("Downloadable Image URL: {}", image_url);
+        }
+
+        let guest_env = GuestEnv::from_stdin(input.clone())
+            .encode()
+            .context("Failed to encode input for boundless proving")?;
+
+        // Deposit to contract
+        // let deposit_amount = U256::from(1e16 as u64); // 0.01eth
+        // let market = self.client.boundless_market.clone();
+        // market.deposit(deposit_amount).await?;
+        // tracing::info!(
+        //     "Successfully deposited {} ETH",
+        //     alloy_primitives::utils::format_units(deposit_amount, "ether")?
+        // );
+
+        // Upload input
+        let mut input_url = self.client.upload_input(&guest_env).await?;
+        tracing::info!("Uploaded input to {}", input_url);
+
+        // If we are not using presigned:
+        if !s3_use_presigned {
+            let s3_path = input_url.as_str().to_string().replace("s3://", "");
+            input_url = Url::parse(&format!("{s3_url}{s3_path}"))?;
+            tracing::info!("Downloadable Input URL: {}", input_url);
+        }
+
+        // move non-Send logic to blocking thread
+        // I had to do this because the executor env builder is not Send
+        let (journal, mcycles_count, total_cycles_approx) = tokio::task::spawn_blocking({
+            let elf = elf.clone(); // clone since we move into thread
+            let input = input.clone();
+            let assumptions = assumptions.clone();
+
+            move || -> anyhow::Result<(Journal, u64, u64)> {
+                let mut env = ExecutorEnvBuilder::default();
+                for assumption in assumptions {
+                    env.add_assumption(assumption);
+                }
+                let env = env.write_slice(&input).build()?;
+
+                let session_info = default_executor().execute(env, &elf)?;
+
+                let total_cycles_approx = session_info
+                    .segments
+                    .iter()
+                    .map(|segment| 1 << segment.po2)
+                    .sum::<u64>();
+                let mcycles_count = total_cycles_approx.div_ceil(1_000_000);
+                tracing::info!(
+                    "Boundless proving session with job id: {} takes {} cycles",
+                    job_id,
+                    total_cycles_approx
+                );
```

Comment:
```suggestion
                tracing::info!(
                    "Boundless proving session with job id: {job_id} takes {total_cycles_approx} cycles"
                );
```

In line with the rest of the PR formatting